### PR TITLE
Rethrow LockedException with full path

### PR DESCRIPTION
**Pull request by PVince81**
Because the path is converted to md5 from the original exception,
rethrow the exception with the correct full path

FYI @icewind1991 @SergioBertolinSG 
- ~~currently returned paths are full paths from the root, for example it says "admin/files/somefile.txt" is locked but it should say "somefile.txt" is locked in the UI. Need to catch the exceptions again higher and strip the path~~

### CODE REVIEW

**Comment by PVince81** [comment]
Considering that the files UI will be using WebDAV later, let's also add the correct message there right now:
- [ ] Specify locked path in WebDAV exception too

**Comment by PVince81** [comment]
The thing that doesn't make me happy is that the ajax endpoints are all over the place, which means I'll need to catch the exception in several places and convert the paths. Oh well... I can't wait to move the files app to WebDAV.

**Comment by icewind1991** [comment]
:+1: makes sense

**Comment by nickvergessen** [comment]
`lockFile` as well, other then that very helpful and much better then the current version.

**Comment by PVince81** [comment]
- [ ] lockFile as well

**Comment by nickvergessen** [comment]
I just saw, that `lockFile()` calls `lockPath()`, so ignore that.

File: lib/private/files/view.php [hunk 1]
```diff
@@ -1665,13 +1665,18 @@ private function getParents($path) {
...
+				$mount->getStorage()->acquireLock(
+					$mount->getInternalPath(
+						$this->getAbsolutePath($path)
+					),
+					$type,
+					$this->lockingProvider
+				);
+			} catch (\OCP\Lock\LockedException $e) {
+				// rethrow with the correct path
+				throw new \OCP\Lock\LockedException($this->getAbsolutePath($path));
```
Comment (nickvergessen) [inline lib/private/files/view.php]: How about using the third parameter of `\Exception`:
```php
 public function __construct($message = "", $code = 0, Exception $previous = null)
```

So throw a new exception, but append the previous one, so we can get to the message in the root point and then can display the lock path and the file path as we want.

Comment (PVince81) [inline lib/private/files/view.php]: Oh cool, I didn't know PHP had this. In Java I'd always use the "reason" parameter to achieve the same result.

Comment (nickvergessen) [inline lib/private/files/view.php]: (you need to tunnel the parameter through `\OCP\Lock\LockedException` thou, because it does not yet have it

**Comment by PVince81** [comment]
Considering that starting with https://github.com/owncloud/core/pull/16884 we will only lock files under "files", then I'll adjust the $path in the exception to strip away "files".

**Comment by PVince81** [comment]
Please review @nickvergessen @SergioBertolinSG 
@icewind1991 I slightly changed the code to properly translate the path, please have a second look

**Comment by icewind1991** [comment]
I would prefer to have absolute paths in the exceptions

**Comment by PVince81** [comment]
Hmmm, but now we only anything under "files" that should be fine.

My problem with absolute paths is that they would be shown as absolute in the web UI and confuse the user.
If we want to change this, we need to catch LockedException again in all ajax/* endpoints and strip away the two first sections.

**Comment by nickvergessen** [comment]
> I would prefer to have absolute paths in the exceptions

I'd disagree, we should only name paths the user can access

**Comment by PVince81** [comment]
Rebased.

@nickvergessen @icewind1991 @DeepDiver1975 another look ?

@SergioBertolinSG this will make the locking error messages more user-friendly.

@jancborchardt I wonder if we should rework "XXX is locked" to "XXX is busy". Because "locked" might give the user the impression that someone else explicitly locked the folder.

File: lib/private/files/view.php [hunk 2]
```diff
@@ -1798,4 +1818,25 @@ protected function shouldLockFile($path) {
...
+	 * if the path was NOT relative to files
+	 *
+	 * @since 8.1.0
+	 */
+	public function getPathRelativeToFiles($absolutePath) {
+		$path = Filesystem::normalizePath($absolutePath);
+		$path = explode('/', trim($path, '/'), 3);
+		// "$user", "files", "path/to/dir"
+		if (!isset($path[2]) || $path[1] !== 'files') {
+			return null;
```
Comment (nickvergessen) [inline lib/private/files/view.php]: Exception? invalid path? :P 
I thought we wanted to stop using obscure return values

**Comment by nickvergessen** [comment]
Looks good to me

**Comment by PVince81** [comment]
@nickvergessen I replaced `return null` with `throw \InvalidArgumentException`.

Unfortunately I noticed too late that I needed to update the `return` tag in PHPDoc.
It will come with the next force-push after Jenkins is done.

**Comment by nickvergessen** [comment]
Testergebnis (3 fehlgeschlagene Tests / ±0)

    Test\Files\View::testGetPathRelativeToFiles.testGetPathRelativeToFiles with data set #0
    Test\Files\View::testGetPathRelativeToFiles.testGetPathRelativeToFiles with data set #1
    Test\Files\View::testGetPathRelativeToFiles.testGetPathRelativeToFiles with data set #2

File: tests/lib/files/view.php [hunk 3]
```diff
@@ -1169,4 +1169,24 @@ public function dataLockPaths() {
 			['', '/testuser/{folder}'],
 		];
 	}
+
+	function pathRelativeToFilesProvider() {
+		return [
+			['x', null],
+			['files', null],
+			['/admin/files', ''],
```
Comment (nickvergessen) [inline tests/lib/files/view.php]: those throw, split into two tests and check for the exception with https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions

File: tests/lib/files/view.php [hunk 4]
```diff
@@ -1169,4 +1169,24 @@ public function dataLockPaths() {
 			['', '/testuser/{folder}'],
 		];
 	}
+
+	function pathRelativeToFilesProvider() {
```
Comment (nickvergessen) [inline tests/lib/files/view.php]: missing visibility

**Comment by PVince81** [comment]
Right, I forgot that I wrote unit tests for that method... Ah the habit of missing tests...

File: tests/lib/files/view.php (hunk 4 above)
Comment (PVince81) [inline tests/lib/files/view.php]: public?

Comment (nickvergessen) [inline tests/lib/files/view.php]: yes

**Comment by PVince81** [comment]
@nickvergessen I found new issues when adding additional test cases in the providers.
Please re-review the function itself too.

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
Looks good, but looks like jenkins needs another branch :(

**Comment by PVince81** [comment]
Jenkins PR: https://github.com/owncloud/core/pull/16952

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Tests passed, please review

**Comment by nickvergessen** [comment]
:+1: 

We shall make use of the method in `shouldLockFile()` too. but this can be done in a new issue.

**Comment by jancborchardt** [comment]
Maybe we should reword it to "is in use"? Both locked and busy do not fully convey what is happening.

**Comment by PVince81** [comment]
Raised ticket for the wording change: https://github.com/owncloud/core/issues/16983
